fairmq/PluginServices.cxx

```diff
@@ -85,3 +86,81 @@ const std::unordered_map<PluginServices::DeviceStateTransition, FairMQDevice::Ev
     {DeviceStateTransition::End,         FairMQDevice::END},
     {DeviceStateTransition::ErrorFound,  FairMQDevice::ERROR_FOUND}
 };
+
+auto PluginServices::ChangeDeviceState(const std::string& controller, const DeviceStateTransition next) -> void
+{
+    // lock_guard<mutex> lock{fDeviceControllerMutex};
+    //
+    // if(!fDeviceController) fDeviceController = controller;
+
+    if(fDeviceController == controller)
+    {
+        fDevice->ChangeState(fkDeviceStateTransitionMap.at(next));
+    }
+    else
+    {
+        throw DeviceControlError{tools::ToString(
+            "Plugin ", controller, " is not allowed to change device states. ",
+            "Currently, plugin ", fDeviceController, " has taken control."
+        )};
+    }
+}
+
+auto PluginServices::TakeDeviceControl(const std::string& controller) -> void
+{
+    lock_guard<mutex> lock{fDeviceControllerMutex};
+
+    if(!fDeviceController)
+    {
+        fDeviceController = controller;
+    }
+    else if(fDeviceController == controller)
+    {
+        // nothing to do
+    }
+    else
+    {
+        throw DeviceControlError{tools::ToString(
+            "Plugin ", controller, " is not allowed to take over control. ",
+            "Currently, plugin ", fDeviceController, " has taken control."
+        )};
+    }
+
+}
+
+auto PluginServices::ReleaseDeviceControl(const std::string& controller) -> void
+{
+    {
+        lock_guard<mutex> lock{fDeviceControllerMutex};
+
+        if(fDeviceController == controller)
+        {
+            fDeviceController = boost::none;
+        }
+        else
+        {
+            throw DeviceControlError{tools::ToString(
+                "Plugin ", controller, " cannot release control because it has not taken over control."
+            )};
+        }
+    }
+
+    fReleaseDeviceControlCondition.notify_one();
+}
+
+auto PluginServices::GetDeviceController() const -> boost::optional<std::string>
+{
+    lock_guard<mutex> lock{fDeviceControllerMutex};
+
+    return fDeviceController;
+}
+
+auto PluginServices::WaitForReleaseDeviceControl() -> void
+{
+    unique_lock<mutex> lock{fDeviceControllerMutex};
+
+    while(GetDeviceController())
```

Comment:
The deadlock happens here - you are locking `fDeviceControllerMutex`, then calling `GetDeviceController`, which tries to lock the same mutex again. I fixed it by just replacing call to `GetDeviceController` with `fDeviceController`. Don't really need the GetDeviceController method, unless you want to use it from somewhere else too.